### PR TITLE
Switch upgradeResourceState to use SDKv2's gRPC method

### PR DESCRIPTION
**Pull request by iwahbe**
This PR switches out the default implementation of `upgradeResourceState` to `schema.GRPCProviderServer.UpgradeResourceState`. 

I explored this PR as part of root causing https://github.com/pulumi/pulumi-github/issues/586. The root cause of https://github.com/pulumi/pulumi-github/issues/586 is https://github.com/pulumi/pulumi-terraform-bridge/issues/1667.

This change is a pure refactor, but it clarifies which behavior is directly from TF and which behavior is part of Pulumi's bridge requests. It also guarantees that if TF changes it's implementation, we will copy the new behavior when we upgrade the linked SDKv2.

### CODE REVIEW

File: pkg/tfshim/sdk-v2/upgrade_state.go [hunk 1]
```diff
@@ -32,61 +34,38 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
...
-		return nil, err
+	var (
+		rawState cty.Value
+		err      error
+	)
+	switch GetInstanceStateStrategy(v2Resource{res}) {
+	// Opt-out of normalization under a flag. Normalization replaces nulls with empty blocks for the state value,
+	// but when processing config the codebase calls NewResourceConfigShimmed which in turn calls
+	// ConfigValueFromHCL2Block which removes empty blocks. This leads to non-empty diffs.
+	case CtyInstanceState:
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/upgrade_state.go:48]: We can start with a PR that removes CtyInstanceState flag from the codebase. org:pulumi search shows that it's' no longer deployed anywhere. I think the use case in AWS was superceded by WithPlanResourceChange.

File: pkg/tfshim/sdk-v2/upgrade_state.go [hunk 2]
```diff
@@ -32,61 +34,38 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
...
+		rawState, err = upgradeResourceStateNonNormalizing(ctx, m, p, int(version), res)
+	default:
+		rawState, err = upgradeResourceStateRPC(ctx, typeName, m, p, res, version)
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/upgrade_state.go:51]: This is a very tedious spot-the-differences exercise hiding here. I've read the before/after code line by line and they seem to be following very closely. 

Before:

```go
json, _, _ := UpgradeFlatmapState(_, _, m, _, _)
json, _ = UpgradeJSONState(json)
RemoveAttributes(_, json, _)
v, _ = JSONMapToStateValue(json, configBlock)
v, _ = CoerceValue(v)
v = schema.NormalizeObjectFromLegacySDK(v, configBlock)
```

After:

```go
func (s *GRPCProviderServer) UpgradeResourceState(
   s.upgradeFlatmapState(ctx, version, req.RawState.Flatmap, res)
   jsonMap, err = s.upgradeJSONState(ctx, version, jsonMap, res)

   RemoveAttributes(ctx, jsonMap, schemaBlock.ImpliedType())
   val, err := JSONMapToStateValue(jsonMap, schemaBlock)
   val, err = schemaBlock.CoerceValue(val)
   val = objchange.NormalizeObjectFromLegacySDK(val, schemaBlock)
   // serialize to msgpack

// deserialize from msgpack        
var newState cty.Value

newState, err := res.ShimInstanceStateFromValue(rawState)
```

It seems that the serialization of cty.Value to msgpack and back is the only addition. That one I think should not be lossy. 

It seems like the previous method was an inlining of the logic in the gRPC provider.

**Comment by t0yv0** [comment]
Have you looked at how this code is covered? 

If we had https://github.com/pulumi/pulumi-terraform-bridge/issues/1736 in place evaluating safety here would have been an easy conversation. I think downstream checks will not test this very well at all, we need good coverage in the bridge. 

I found this:
https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tests/regress_923_test.go#L29

But are the more tests covering the upgrade specifically?

**Comment by iwahbe** [comment]
> I found this: https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tests/regress_923_test.go#L29
> 
> But are the more tests covering the upgrade specifically?

There are not. I tried turning off `upgradeResourceState` (making it a no-op) and that was the only test that failed.

File: pkg/tfshim/sdk-v2/upgrade_state.go (hunk 1 above)
Comment (iwahbe) [reply pkg/tfshim/sdk-v2/upgrade_state.go]: Done. That makes this PR simpler.

File: pkg/tfshim/sdk-v2/upgrade_state.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfshim/sdk-v2/upgrade_state.go]: That is my understanding, which I why this is a no-op refactor, not a bug fix.

**Comment by t0yv0** [comment]
Ahh yes, very unfortunate.. Well it seems like we should consider throwing some more simple tests specifically designed at covering upgrading resources basics before changing the code further. Especially if this is mostly a forward-looking refactor and we do not have customers waiting on a change of behavior here.

File: pkg/tfshim/sdk-v2/upgrade_state.go [hunk 3]
```diff
@@ -32,56 +34,53 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
...
-	}
+	// Now, we perform the UpgradeResourceState operation by re-implementing TF's UpgradeResourceState.
 
-	// Next, migrate the JSON state up to the current version.
-	json, err = schema.UpgradeJSONState(ctx, version, json, res, p.Meta())
+	resp, err := schema.NewGRPCProviderServer(p).
+		UpgradeResourceState(ctx, &tfprotov5.UpgradeResourceStateRequest{
+			TypeName: typeName,
+			Version:  version,
+			RawState: &tfprotov5.RawState{Flatmap: m},
```
Comment (VenelinMartinov) [inline pkg/tfshim/sdk-v2/upgrade_state.go:46]: I believe you said the flatmap representation is problematic, right?

Comment (VenelinMartinov) [reply pkg/tfshim/sdk-v2/upgrade_state.go:46]: Under the hood this calls UpgradeFlatmapState

Comment (iwahbe) [reply pkg/tfshim/sdk-v2/upgrade_state.go]: It is, and that is fixed in [c9ea19c](https://github.com/pulumi/pulumi-terraform-bridge/pull/1735/commits/c9ea19c2fd466d7f49c98724de0f7c5dd7ee2f50).

File: pkg/tfshim/sdk-v2/upgrade_state.go [hunk 4]
```diff
@@ -35,13 +39,18 @@ func upgradeResourceState(ctx context.Context, typeName string, p *schema.Provid
 		version, hasVersion = v, true
 	}
 
-	// Now, we perform the UpgradeResourceState operation by re-implementing TF's UpgradeResourceState.
+	json, err := ctyjson.Marshal(instanceState.RawState, res.CoreConfigSchema().ImpliedType())
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/upgrade_state.go:42]: Eyes on this line please, I think  res.CoreConfigSchema().ImpliedType() comes from the current provider schema but instanceState.RawState may be on prev provider schema in the case of actual upgrade. We could try inferring the type from the value. I think it's about tradeoffs both approaches slightly problematic.

Comment (iwahbe) [reply pkg/tfshim/sdk-v2/upgrade_state.go:52]: We should be able to change that to 
```go
	json, err := ctyjson.Marshal(instanceState.RawState, res.RawState.Type())
```

I'll add a test where we do a more substantial upgrade to verify that this works as expected.

Comment (iwahbe) [reply pkg/tfshim/sdk-v2/upgrade_state.go]: It didn't work without the change. Good catch! I have added the fix and a test to go with it.

File: pkg/tfshim/sdk-v2/upgrade_state.go [hunk 5]
```diff
@@ -6,82 +6,118 @@ import (
...
 			return nil, fmt.Errorf("unexpected type %T for schema_version", versionValue)
 		}
 		v, err := strconv.ParseInt(versionString, 0, 32)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to parse schema_version: %w", err)
 		}
-		version, hasVersion = int(v), true
+		version, hasVersion = v, true
 	}
```
Comment (VenelinMartinov) [inline pkg/tfshim/sdk-v2/upgrade_state.go:34]: I don't see a check for downgrades here - pretty sure TF doesn't allow downgrades, see https://github.com/pulumi/pulumi-terraform-bridge/pull/1998#discussion_r1607158351

**Comment by iwahbe** [comment]
Closing in favor of https://github.com/pulumi/pulumi-terraform-bridge/pull/2005.

**Pull request closed without merging**